### ISSUE TITLE
No `HydrateFallback` element provided 

### ISSUE THREAD

**Issue by carlos-lopez-SaRa** [remix-run/react-router#12563]
### I'm using React Router as a...

library

### Reproduction

I updated to V7, should not be seeing this warning since I'm not using the library as framework.

### System Info

```shell
System:
    OS: Linux 5.15 Ubuntu 22.04.5 LTS 22.04.5 LTS (Jammy Jellyfish)
    CPU: (32) x64 13th Gen Intel(R) Core(TM) i9-13900KF
    Memory: 9.95 GB / 15.53 GB
    Container: Yes
    Shell: 5.8.1 - /usr/bin/zsh
  Binaries:
    Node: 22.12.0 - ~/.nvm/versions/node/v22.12.0/bin/node
    npm: 10.9.2 - ~/.nvm/versions/node/v22.12.0/bin/npm
  Browsers:
    Chrome: 131.0.6778.85
  npmPackages:
    react-router: ^7.0.1 => 7.0.2 
    vite: ^6.0.1 => 6.0.3
```

### Used Package Manager

npm

### Expected Behavior

As I'm using React Router as a Library, there should be no Hydration warnings.

### Actual Behavior

![image](https://github.com/user-attachments/assets/093d306c-5176-4c4e-9a90-ca05dade7b41)

**Comment by timdorr** [comment]
Can you provide a reproduction on StackBlitz?

**Comment by rothsandro** [comment]
This happens when `lazy` is used to render a component. AFAIK this wasn't necessary in v6 and isn't mentioned in the upgrade guide (or did I miss it?).

Repro: https://stackblitz.com/edit/vitejs-vite-6xzkmjng

**Comment by carlos-lopez-SaRa** [comment]
I do have 1 component that is not in a route that is using Lazy, basically being loaded when a drawer opens.

**Comment by brookslybrand** [comment]
@rothsandro it is mentioned in the upgrade guide: https://reactrouter.com/upgrading/v6#v7_partialhydration

@carlos-lopez-SaRa this isn't specific to framework mode, this warning is legitimate and documented in the upgrade guide. I'm going to close this, but please reach out/respond if you have any further questions.

**Comment by carlos-lopez-SaRa** [comment]
I see it now, though it's in a different section: https://reactrouter.com/upgrading/v6#v7_starttransition

**Comment by rothsandro** [comment]
@brookslybrand The section is talking about SSR so I wouldn't expect this to apply to client-side rendered apps with router in library mode. But if it's the expected behavior we'll just add a `hydrateFallback`.

**Comment by carlos-lopez-SaRa** [comment]
Can I add an empty fallback?

**Comment by brookslybrand** [comment]
@carlos-lopez-SaRa yes, I was referring to the `HydrateFallback`. The language in the upgrade doc is a bit confusing (thanks @rothsandro for also pointing that out), I don't believe it's just for SSR (confirming with a team member). I'll update the language once I get confirmation, but it's needed in cases where you're using `lazy` as well (even for a pure SPA)

> Can I add an empty fallback?

Yes, you can provide an empty fallback if you want to silence the error, or a loading spinner or whatever you want

```
HydrateFallback: () => null,

// or 

hydrateFallbackElement: <></>,
```

 The `HydrateFallback` is shown before your route module finishes loading. If you're fine with it being completely blank beforehand, that's fine. Definitely not need to move away from `createBrowserRouter`

**Comment by mtr1990** [comment]
Hi @brookslybrand ,

I have looked at the documentation many times but
`hydrateFallbackElement` and `HydrateFallback` but it still doesn't work. Can you help?

I expected loading to show up when changing routers
https://stackblitz.com/edit/github-cv1g5y7j?file=src%2Fmain.tsx,src%2FApp.tsx

**Comment by brookslybrand** [comment]
> Hi @brookslybrand ,
> 
> I have looked at the documentation many times but `hydrateFallbackElement` and `HydrateFallback` but it still doesn't work. Can you help?
> 
> I expected loading to show up when changing routers https://stackblitz.com/edit/github-cv1g5y7j?file=src%2Fmain.tsx,src%2FApp.tsx

Yeah, I think I was misunderstanding/misleading -- `HydrateFallback` is only for the initial render, not for all navigations to lazy loaded routes. It's the fallback while the app is "hydrating", which is a bit of a confusing name within the library mode.

@brophdawg11 I went ahead and merged #12594, but honestly the language might not be quite right if you want to check it and correct at all

**Comment by mtr1990** [comment]
@brookslybrand 

Is there any solution in `react-router ` currently that can solve this problem?

My expectation is that the loader appears when the router redirects.

I am currently using this method. It works as expected. But in your documentation it is mentioned that this is the wrong method to use.

```tsx
const MyPage = lazy(() => import('src/pages/my-page'));

function SuspenseOutlet() {
  const pathname = usePathname();
  return (
    <Suspense key={pathname} fallback={<LoadingScreen />}>
      <Outlet />
    </Suspense>
  );
}

  {
    path: '/',
    element: (
        <MainLayout>
          <SuspenseOutlet />
        </MainLayout>
    ),
    children: [
      { index: true, element: <MyPage /> },
      ]
...
```
I have read the document many times but am still very vague about this.

I can do this simply in next.js with the loading feature
https://nextjs.org/docs/app/building-your-application/routing/loading-ui-and-streaming#instant-loading-states


Also related to:
https://github.com/remix-run/react-router/issues/12474

**Comment by dbaltor** [comment]
I'm using using `react-router v7` as a library just like @carlos-lopez-SaRa and I don't understand either why we need to specify `HydrateFallback` . I still got this warning despite having declared a fallback in my react code:
```
export function SupplierRisk() {
  const data = useLoaderData<{ suppliers: Supplier[] }>();
  return (
    <Suspense fallback={<div>Loading...</div>}>
      <Await resolve={data.suppliers}>
        {(suppliers) => (
          <SuppliersProvider suppliers={suppliers}>
            // omitted...
          </SuppliersProvider>
        )}
      </Await>
    </Suspense>
  );
}
```
I ended up adding `HydrateFallback: () => null,` to my routes declaration as suggested by @brookslybrand. 
Is there a better way to deal with this?
I'd appreciate any advice.

**Comment by waynesbrain** [comment]
I am using react-router in "data mode" and soon after using async route `loader` functions I started getting this error, so I guess the `HydrateFallback` is just required anytime you make an async route. Probably would be a good thing to mention in the docs here - https://reactrouter.com/start/data/data-loading

```ts
{
      path: "/thing/:id",
      Component: ThingEditor,
      HydrateFallback: () => null,
      loader: loadThingData,
      children: [...],
}
```

**Comment by julienw** [comment]
Hey folks, just a note that this happens also even without a lazy. It may happen as soon as your loader returns a Promise? (which `lazy` does as well, I believe).

I find this issue a bit annoying, but in the same time it gently forces us to make a decision about providing some feedback to the user while the app is loading.

**Comment by wimbarelds** [comment]
> Hey folks, just a note that this happens also even without a lazy. It may happen as soon as your loader returns a Promise? (which `lazy` does as well, I believe).
> 
> I find this issue a bit annoying, but in the same time it gently forces us to make a decision about providing some feedback to the user while the app is loading.

It happens as soon as you have a loader, even if that loader is synchronous and returns static data.
